lib/btree.js

```diff
@@ -0,0 +1,240 @@
+  }
+}
+
+const joinNodes = (parent, firstNodeIndex, secondNodeIndex) => {
+  const firstNode = parent[firstNodeIndex].child;
+  const secondNode = parent[secondNodeIndex].child;
+  const mid = parent.splice(firstNodeIndex, 1)[0];
+  mid.child = firstNode.pop().child;
+  secondNode.unshift(mid);
+  secondNode.splice(0, 0, ...firstNode);
+  return secondNode;
+};
+
+const growChild = (parent, childIndex) => {
+  const parentElement = parent[childIndex];
+  const node = parentElement.child;
+  const minDegree = node.length;
+  if (childIndex > 0) { // If we have left neighbor
+    const leftNeighbor = parent[childIndex - 1].child;
+    if (leftNeighbor.length > minDegree) {
+      const extractedElement =
+        leftNeighbor.splice(leftNeighbor.length - 2, 1)[0];
+      const insertedElement = new Element(
+        parent[childIndex - 1].value,
+        parent[childIndex - 1].data,
+        leftNeighbor[leftNeighbor.length - 1].child
+      );
+      leftNeighbor[leftNeighbor.length - 1].child = extractedElement.child;
+      extractedElement.child = leftNeighbor;
+      parent.splice(childIndex - 1, 1, extractedElement);
+      node.unshift(insertedElement);
+      return node;
+    }
+  } else if (childIndex < parent.length - 1) { // If we have right neighbor
+    const rightNeighbor = parent[childIndex + 1].child;
+    if (rightNeighbor.length > minDegree) {
+      const extractedElement = rightNeighbor.shift();
+      const insertedElement = new Element(
+        parentElement.value,
+        parentElement.data,
+        node[node.length - 1].child
+      );
+      node[node.length - 1].child = extractedElement.child;
+      extractedElement.child = node;
+      parent.splice(childIndex, 1, extractedElement);
+      node.splice(node.length - 1, 0, insertedElement);
+      return node;
+    } else return joinNodes(parent, childIndex, childIndex + 1);
+  }
+  return joinNodes(parent, childIndex - 1, childIndex);
+};
+
+const extractUpperLimit = (node, minDegree) => {
+  let currNode = node;
+  while (!(isLeaf(currNode))) {
+    let nextNode = currNode[0].child;
+    if (nextNode.length === minDegree) {
+      nextNode = growChild(currNode, 0);
+    }
+    currNode = nextNode;
+  }
+  return currNode.shift();
+};
+
+const extractLowerLimit = (node, minDegree) => {
+  let currNode = node;
+  while (!(isLeaf(currNode))) {
+    let nextNode = currNode[currNode.length - 1].child;
+    if (nextNode.length === minDegree) {
+      nextNode = growChild(currNode, currNode.length - 1);
+    }
+    currNode = nextNode;
+  }
+  const test = currNode.splice(currNode.length - 2, 1)[0];
+  return test;
+};
+
+const deleteElement = (node, elementIndex, minDegree) => {
+  const element = node[elementIndex];
+  const deletedData = element.data;
+  const leftChild = element.child;
+  const rightChild = node[elementIndex + 1].child;
+  if (leftChild.length > minDegree) {
+    const lowerLimit = extractLowerLimit(leftChild, minDegree);
```

Comment:
I'm not sure, but shouldn't you search for the highest element in the left child and lowest in the right? (here it is reversed) Or maybe it's the function name that are confusing?

Reply:
I thought so: the upper limit of an element is the smallest element in the right node, and the lower limit is the smallest element in the left node. Is it more correct to rename?

Reply:
The right node contains all elements bigger than this one, and you find the smallest element in that node, therefore, it'd be better to call it by exactly what it does - finds the smallest element (and same goes for the other function but reverted).